api/src/main/java/org/openmrs/module/appointments/dao/impl/AppointmentDaoImpl.java

```diff
@@ -61,6 +61,25 @@ public List<Appointment> getAllAppointmentsReminder(String hours) {
         return criteria.list();
     }
 
+    @Override
+    public List<Appointment> getAllAppointments(Date forDate, String status) {
+        Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Appointment.class);
+        criteria.add(Restrictions.eq("voided", false));
+        criteria.createAlias("patient", "patient");
+        criteria.add(Restrictions.eq("patient.voided", false));
+        criteria.add(Restrictions.eq("patient.personVoided", false));
+        if (forDate != null) {
+            Date maxDate = new Date(forDate.getTime() + TimeUnit.DAYS.toMillis(1));
+            criteria.add(Restrictions.ge("startDateTime", forDate));
+            criteria.add(Restrictions.lt("endDateTime", maxDate));
```

Comment:
There are a number of things here that seem wrong to me, but maybe I'm not fully understanding what this method aims to accomplish.

1. Your restrictions are not operating on the same column.  If the intention is to find all appointments that are optionally limited by date, I would think you'd want to only restrict on `startDatetime` (or at the very least you'd want to limit the `startDatetime` to both be within the passed date.
2. The way in which you are restricting the data also seems wrong.  Assuming what you want is all appointments on the given date, if someone passes in a `forDate` with a time component, the first criteria will not include appointments before the given time on that date.  And the second criteria will include appointments for the _next_ date, up until the time component of the original `forDate`.  You should first zero-out the time-component of `forDate` before using it.
3. The approach to get "one day later" also seems incorrect (by adding a day's worth of millis).  For example, what would happen on daylight savings dates where an hour is gained or lost?  You would not end up with the next day at the same time.  A better approach would be to use standard date/time functions (or Calendar) to add a day.

Reply:
@mseaton getAllAppointments is really not our addition. Please see https://github.com/Bahmni/openmrs-module-appointments/blob/master/api/src/main/java/org/openmrs/module/appointments/dao/impl/AppointmentDaoImpl.java#L34. The reviews are good and are worth addressing. They, however, should not block accepting the changes that were added by members outside of Bahmni team. I propose a comprehensive review of the module with a view of improving the legacy code if that is agreeable. @angshu what's your thoughts on this?

Reply:
@ojwanganto I see what you mean.  Still, it would probably be better if we could not duplicate this code across multiple DAO methods, as if it needs to be changed then it will need to be done consistently in multiple places.  Why not change the existing DAO to take in an optional status, and limit by this if it is non-null?  Then your PR makes clear that it is not introducing any code that has potentially questionable logic.

@angshu also interested in your opinion on my original comments on - what is existing functionality in the getAllAppointment method implementation.

Reply:
Agree with @mseaton about changing the existing service to take an optional status instead of creating a new method.  If we need to preserve the existing message signature, we can keep the existing signature but have it delegate to the one that takes status, ie:

```
public List<Appointment> getAllAppointments(Date forDate) {
    return getAllAppointments(forDate, null)
}
```

If we do this, we can merge, and we can address the concerns @mseaton had about the original method separately.
.